packages/origin-247-certificate/src/offchain-certificate/offchain-certificate.service.ts

```diff
@@ -168,25 +190,25 @@ export class OffChainCertificateService<T = null> {
         await this.validateBatchClaim(commands);
 
         for (const command of commands) {
-            await this.claim(command);
+            await this.claimWithoutValidation(command);
         }
     }
 
     public async batchTransfer(commands: ITransferCommand[]): Promise<void> {
         await this.validateBatchTransfer(commands);
 
         for (const command of commands) {
-            await this.transfer(command);
+            await this.transferWithoutValidation(command);
         }
     }
 
     private async validateBatchIssue(commands: IIssueCommand<T>[]): Promise<void> {
         try {
-            commands.forEach(async (c) => {
-                CertificateAggregate.fromEvents([
-                    CertificateIssuedEvent.createNew(await this.generateInternalCertificateId(), c)
-                ]);
+            commands.forEach((c) => {
+                CertificateAggregate.fromEvents([CertificateIssuedEvent.createNew(-1, c)]);
             });
+
+            await issueBatchCommandValidator(commands);
```

Comment:
You should validate first, then apply to aggregate. Basically IMHO it should be validated as soon as possible when provided by user. I would validate this in the first line of the API function, otherwise IMHO it is too obscure, that something is validated. You can even use parameter decorator for issue/transfer/claim functions. It is all to ensure that you are doing anything, any business logic on correct data, and not some strange invalid stuff